<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 1: World Development Statistics

--- 
# Part 1

Part 1 requires knowledge of basic Python.

---

## Problem Statement

How do life expectancy, health spending, and population trends differ across regions (e.g., Asia, Europe, Africa)?

### Contents:
- [Background](#Background)
- [Data Import & Cleaning](#Data-Import-and-Cleaning)
- [Exploratory Data Analysis](#Exploratory-Data-Analysis)
- [Data Visualization](#Visualize-the-Data)
- [Conclusions and Recommendations](#Conclusions-and-Recommendations)

## Background

For our first project, we're going to take a look at World Development Statistics from [Gapminder](https://www.gapminder.org/about/), an independent Swedish foundation that aims to make data about the world more accessible and reliable. A good introduction on Gapminder is this [Ted Talk](https://www.ted.com/talks/hans_rosling_the_best_stats_you_ve_ever_seen) from Hans Rosling, which also shows how effective data visualization can be for your audience.

**To-Do:** *Fill out this cell (or edit the above cell) with any other background or information that is necessary for your problem statement.*

### Choose your Data

There are 3 datasets included in the [`data`](./data/) folder for this project. You are required to pick **at least two** of these to complete your analysis. Feel free to use more than two if you would like, or add other relevant datasets you find online. 

**HINT: There are more datasets available on the Gapminder website**

* [`population.csv`](./data/population.csv): Population by Country
* [`life_expectancy.csv`](./data/life_expectancy.csv): Life Expectancy by Country
* [`gni_per_cap_atlas_method_con2021.csv`](./data/gni_per_cap_atlas_method_con2021.csv): Gross National Income (GNI) per capita in current US dollars

I will be using the following datasets:

1. population.csv
2. life_expectancy.csv
3. gov_health_spending.csv

### Outside Research

Based on your problem statement and your chosen datasets, spend some time doing outside research on state policies or additional information that might be relevant. Summarize your findings below. If you bring in any outside tables or charts, make sure you are explicit about having borrowed them. If you quote any text, make sure that it renders as being quoted. **Make sure that you cite your sources.**

**To-Do:** *Fill out this cell with outside research or any additional background information that will support your analysis.*

### Coding Challenges

1. Manually calculate mean:

    Write a function that takes in values and returns the mean of the values. Create a list of numbers that you test on your function to check to make sure your function works!
    
    *Note*: Do not use any mean methods built-in to any Python libraries to do this! This should be done without importing any additional libraries.

In [374]:
# Code:
nums = [14, 98, 72, 53, 9, 0, 12, 77, 25, 33]

def calculate_mean(nums):
    total = 0
    for n in nums:
        total += n
    return total / len(nums)

In [375]:
calculate_mean(nums)

39.3

The mean of nums is 39.3

2. Manually calculate standard deviation:

    The formula for standard deviation is below:

    $$\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^n(x_i - \mu)^2}$$

    Where $x_i$ represents each value in the dataset, $\mu$ represents the mean of all values in the dataset and $n$ represents the number of values in the dataset.

    Write a function that takes in values and returns the standard deviation of the values using the formula above. Hint: use the function you wrote above to calculate the mean! Use the list of numbers you created above to test on your function.
    
    *Note*: Do not use any standard deviation methods built-in to any Python libraries to do this! This should be done without importing any additional libraries.

In [376]:
# Code:

def calc_std_dev(data):
    # find mean of data
    mean = calculate_mean(data)
    # find squared difference between values
    squared_diff = [(x - mean) ** 2 for x in data]
    # find squared difference mean
    squared_mean = calculate_mean(squared_diff)
    # find square root of squared_mean
    std_dev = squared_mean ** 0.5
    return std_dev

calc_std_dev(nums)

31.962634434601917

The standard deviation of nums is 31.96

3. Data cleaning function:
    
    Write a function that takes in a string consisting a number and an abbreviated place value (ex. '5M', '45K'), and converts that string to an integer. For example, inputting '5M' in your function should return 5000000, '45K' should return 45000, etc. Make sure to test your function to make sure it works!

You will use these functions later on in the project!

In [377]:
# Code:
cipher = {'k':'000', 'K':'000', 'm':'000000', 'M':'000000', 'b':'000000000', 'B':'000000000'}
string = '6b'

def str_to_num(string, cipher):
    new_string = ''
    for char in string:
        if char in cipher:
            new_string += cipher[char]
        else:
            new_string += char
    return int(new_string)

str_to_num(string, cipher)
            

6000000000

--- 
# Part 2

Part 2 requires knowledge of Pandas, EDA, data cleaning, and data visualization.

---

*All libraries used should be added here*

## Data Import and Cleaning

In [378]:
# Imports:

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

population = pd.read_csv('../data/population.csv')
life_exp = pd.read_csv('../data/life_expectancy.csv')
gov_health_spending = pd.read_csv('../data/government_health_spending_of_total_gov_spending_percent.csv')

### Data Import & Cleaning

Import the datasets that you selected for this project and go through the following steps at a minimum. You are welcome to do further cleaning as you feel necessary:
1. Display the data: print the first 5 rows of each dataframe to your Jupyter notebook.
2. Check for missing values.
3. Check for any obvious issues with the observations (keep in mind reasonable values for each column).
4. Fix any errors you identified in steps 2-3.
5. Display the data types of each feature.
6. Fix any incorrect data types found in step 5.
    - Fix any individual values preventing other columns from being the appropriate type.
    - If your dataset has a column of abbreviated numbers (ex. '3M', '45K', etc.), use the function you wrote in Part 1 (coding challenges, number 3) to convert this to integers! *Hint*: use `.map()` or `.apply()`.
7. Rename Columns.
    - Column names should be all lowercase.
    - Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`).
    - Column names should be unique and informative.
8. Drop unnecessary rows (if needed).
9. Merge dataframes that can be merged.
10. Perform any additional cleaning that you feel is necessary.
11. Save your cleaned and merged dataframes as csv files.

## 1. Display the data

In [379]:
population.head()

country   1800   1801   1802   1803   1804   1805   1806  \
0           Afghanistan  3.28M  3.28M  3.28M  3.28M  3.28M  3.28M  3.28M   
1                Angola  1.57M  1.57M  1.57M  1.57M  1.57M  1.57M  1.57M   
2               Albania   400k   402k   404k   405k   407k   409k   411k   
3               Andorra   2650   2650   2650   2650   2650   2650   2650   
4  United Arab Emirates  40.2k  40.2k  40.2k  40.2k  40.2k  40.2k  40.2k   

    1807   1808  ...   2091   2092   2093   2094   2095   2096   2097   2098  \
0  3.28M  3.28M  ...   108M   108M   109M   109M   109M   110M   110M   110M   
1  1.57M  1.57M  ...   125M   126M   127M   128M   129M   130M   131M   131M   
2   413k   414k  ...  1.35M  1.32M  1.29M  1.26M  1.23M  1.21M  1.18M  1.15M   
3   2650   2650  ...  62.5k  62.2k  61.9k  61.7k  61.4k  61.2k  60.9k  60.7k   
4  40.2k  40.2k  ...  13.5M  13.5M  13.6M  13.7M  13.8M  13.8M  13.9M    14M   

    2099   2100  
0   111M   111M  
1   132M   133M  
2  1.12M   1.1M  
3  60.5k  60.2k  
4    14M  14.1M  

[5 rows x 302 columns]

In [380]:
life_exp.head()

country  1800  1801  1802  1803  1804  1805  1806  1807  1808  \
0           Afghanistan  28.2  28.2  28.2  28.2  28.2  28.2  28.1  28.1  28.1   
1                Angola  27.0  27.0  27.0  27.0  27.0  27.0  27.0  27.0  27.0   
2               Albania  35.4  35.4  35.4  35.4  35.4  35.4  35.4  35.4  35.4   
3               Andorra   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  United Arab Emirates  30.7  30.7  30.7  30.7  30.7  30.7  30.7  30.7  30.7   

   ...  2091  2092  2093  2094  2095  2096  2097  2098  2099  2100  
0  ...  75.5  75.7  75.8  76.0  76.1  76.2  76.4  76.5  76.6  76.8  
1  ...  78.8  79.0  79.1  79.2  79.3  79.5  79.6  79.7  79.9  80.0  
2  ...  87.4  87.5  87.6  87.7  87.8  87.9  88.0  88.2  88.3  88.4  
3  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4  ...  82.4  82.5  82.6  82.7  82.8  82.9  83.0  83.1  83.2  83.3  

[5 rows x 302 columns]

In [381]:
gov_health_spending.head()

country   1995   1996   1997   1998   1999   2000   2001   2002   2003  \
0  Afghanistan    NaN    NaN    NaN    NaN    NaN    NaN    NaN   1.48   1.48   
1       Angola   5.00   2.68   3.57   3.15   1.76   3.26   6.06   3.74   4.83   
2      Albania   5.26   6.34   6.47   6.10   7.18   7.03   7.24   7.32   7.64   
3      Andorra  23.60  23.80  23.20  28.70  20.80  19.10  19.20  20.00  22.00   
4          UAE   8.09   7.13   8.76   8.00   8.01   7.64   7.73   7.98   8.35   

    2004   2005   2006   2007   2008   2009   2010  
0   1.48   1.48   1.48   1.48   1.48   1.58   1.59  
1   4.12   4.38   6.06   5.75   6.40  10.10   7.18  
2   9.23   9.79   9.05   8.46   8.21   8.42   8.42  
3  22.70  22.00  22.80  21.30  21.30  21.30  21.30  
4   8.21   8.70   8.95   8.93   8.85   8.76   8.79

In [382]:
# Since my dataset for Government Health Spending only spans the years 1995-2010, those are the years we'll look at.

population = population[['country', '1995', '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004', '2005', '2006', '2007', '2008', '2009', '2010']]
life_exp = life_exp[['country', '1995', '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004', '2005', '2006', '2007', '2008', '2009', '2010']]

In [383]:
population.head()

country   1995   1996   1997   1998   1999   2000   2001  \
0           Afghanistan  16.4M  17.1M  17.8M  18.5M  19.3M  19.5M  19.7M   
1                Angola  13.9M  14.4M  14.9M  15.4M  15.9M  16.4M  16.9M   
2               Albania  3.28M  3.27M  3.25M  3.23M  3.21M  3.18M  3.15M   
3               Andorra  62.9k  64.1k  64.7k  65.2k  65.7k  66.1k  67.8k   
4  United Arab Emirates  2.43M  2.57M  2.75M  2.92M   3.1M  3.28M  3.45M   

    2002   2003   2004   2005   2006   2007   2008   2009   2010  
0    21M  22.6M  23.6M  24.4M  25.4M  25.9M  26.4M  27.4M  28.2M  
1  17.5M  18.1M  18.8M  19.5M  20.2M  20.9M  21.7M  22.5M  23.4M  
2  3.12M  3.09M  3.06M  3.03M     3M  2.98M  2.95M  2.93M  2.91M  
3  70.8k  73.9k  76.9k  79.8k  80.2k  78.2k  76.1k  73.9k  71.5k  
4  3.63M  3.81M  3.99M  4.28M   4.9M  5.87M  6.99M  7.99M  8.48M

In [384]:
life_exp.head()

country  1995  1996  1997  1998  1999  2000  2001  2002  2003  \
0           Afghanistan  54.3  54.7  54.5  53.3  54.7  54.7  54.8  55.5  56.5   
1                Angola  51.2  51.7  51.6  50.6  51.9  52.8  53.4  54.5  55.1   
2               Albania  74.6  74.5  72.9  74.8  75.1  75.4  76.0  75.9  75.6   
3               Andorra  79.8  80.0  80.2  80.4  80.6  80.8  80.9  81.1  81.2   
4  United Arab Emirates  68.8  68.9  69.0  69.2  69.2  69.1  69.2  69.4  69.3   

   2004  2005  2006  2007  2008  2009  2010  
0  57.1  57.6  58.0  58.5  59.2  59.9  60.5  
1  55.5  56.4  57.0  58.0  58.8  59.5  60.2  
2  75.8  76.2  76.9  77.5  77.6  78.0  78.1  
3  81.3  81.4  81.5  81.7  81.8  81.8  81.8  
4  69.1  69.2  69.5  70.0  70.4  70.6  70.8

## 2. Check for missing values

In [385]:
# Exploring data for missing values

population.isna().sum()

country    0
1995       0
1996       0
1997       0
1998       0
1999       0
2000       0
2001       1
2002       1
2003       1
2004       1
2005       1
2006       1
2007       1
2008       1
2009       1
2010       1
dtype: int64

In [386]:
life_exp.isna().sum()

country    0
1995       0
1996       0
1997       0
1998       0
1999       0
2000       0
2001       0
2002       0
2003       0
2004       0
2005       0
2006       0
2007       0
2008       0
2009       0
2010       0
dtype: int64

No null values in this dataset! Yay!

In [387]:
# Checking our Government Health Spending dataset for missing values

gov_health_spending.isna().sum()

country    0
1995       3
1996       2
1997       2
1998       1
1999       1
2000       1
2001       1
2002       2
2003       2
2004       2
2005       2
2006       2
2007       2
2008       2
2009       2
2010       5
dtype: int64

## 3. Check for any obvious issues with the observations.

In [388]:
life_exp.describe()

1995        1996        1997        1998        1999        2000  \
count  195.000000  195.000000  195.000000  195.000000  195.000000  195.000000   
mean    66.590769   66.789744   66.900000   67.016410   67.169231   67.489231   
std      8.971180    9.123059    9.291832    9.378174    9.469483    9.487188   
min     43.400000   43.000000   43.000000   43.800000   43.800000   43.800000   
25%     60.450000   60.400000   60.150000   60.250000   60.200000   60.800000   
50%     69.000000   69.500000   69.800000   69.900000   70.200000   70.300000   
75%     73.400000   73.500000   73.850000   74.100000   74.400000   74.700000   
max     80.700000   80.900000   81.100000   81.300000   81.500000   81.600000   

             2001        2002        2003        2004        2005        2006  \
count  195.000000  195.000000  195.000000  195.000000  195.000000  195.000000   
mean    67.770769   67.963077   68.207692   68.470769   68.818462   69.181538   
std      9.492526    9.474576    9.433451    9.355224    9.246262    9.049649   
min     43.900000   44.200000   43.900000   43.100000   43.300000   44.100000   
25%     61.150000   61.450000   61.950000   62.150000   62.200000   62.350000   
50%     70.600000   70.700000   71.000000   71.100000   71.200000   71.200000   
75%     75.000000   75.050000   75.250000   75.450000   75.600000   75.900000   
max     81.900000   82.200000   82.300000   82.500000   82.500000   82.800000   

             2007        2008        2009        2010  
count  195.000000  195.000000  195.000000  195.000000  
mean    69.542564   69.856410   70.205128   70.445128  
std      8.847965    8.656599    8.435973    8.704035  
min     44.900000   45.500000   46.400000   32.500000  
25%     62.800000   63.250000   63.900000   64.200000  
50%     71.600000   71.900000   72.100000   72.300000  
75%     75.950000   75.950000   76.300000   76.500000  
max     82.900000   83.100000   83.300000   83.300000

In the year 2010, we can see the minimum life expectancy drops significantly and suddenly.

In [389]:
life_exp[life_exp['2010'] == life_exp['2010'].min()]

country  1995  1996  1997  1998  1999  2000  2001  2002  2003  2004  2005  \
75   Haiti  55.4  55.7  56.1  56.5  57.0  57.4  57.8  58.2  58.5  57.5  59.0   

    2006  2007  2008  2009  2010  
75  59.5  60.0  60.2  60.7  32.5

 The minimum value for life expectency belongs to the Haiti index. A quick google search of Haiti 2010 reveals a catastrophic earthquake that killed over 200,000 people, explaining the outlier.

[Link: Haiti 2010 Earthquake](https://www.britannica.com/event/2010-Haiti-earthquake)

In [390]:
gov_health_spending.describe()

1995        1996        1997        1998        1999        2000  \
count  189.000000  190.000000  190.000000  191.000000  191.000000  191.000000   
mean    10.053757   10.184368   10.369258   10.422932   10.341513   10.491832   
std      3.966658    4.125900    4.142478    4.248769    4.198102    4.362539   
min      1.480000    0.130000    0.130000    0.100000    0.100000    0.100000   
25%      7.160000    7.077500    7.635000    7.790000    7.385000    7.445000   
50%      9.850000    9.955000    9.995000    9.930000    9.950000    9.890000   
75%     12.400000   12.950000   12.975000   13.400000   12.950000   13.200000   
max     23.600000   23.800000   24.600000   28.700000   23.100000   21.700000   

             2001        2002        2003        2004        2005        2006  \
count  191.000000  190.000000  190.000000  190.000000  190.000000  190.000000   
mean    10.934251   10.909526   11.322053   11.335105   11.441321   11.519737   
std      5.110299    4.839801    4.667015    4.876756    4.966433    4.700775   
min      0.150000    0.100000    1.140000    1.250000    0.771000    1.200000   
25%      8.055000    7.947500    8.210000    8.240000    8.065000    8.087500   
50%     10.300000   10.550000   10.800000   10.650000   11.300000   11.300000   
75%     13.350000   13.375000   13.900000   14.100000   14.550000   14.900000   
max     42.400000   41.200000   35.500000   35.500000   38.300000   25.300000   

             2007        2008        2009        2010  
count  190.000000  190.000000  190.000000  187.000000  
mean    11.465137   11.353216   11.434584   11.427759  
std      4.682573    4.584682    4.703264    4.722166  
min      0.926000    0.911000    0.971000    0.971000  
25%      8.225000    7.720000    7.922500    8.040000  
50%     11.150000   11.500000   11.450000   11.300000  
75%     14.675000   14.550000   14.700000   14.450000  
max     24.600000   26.100000   30.600000   29.000000

No extreme outliers in the data, other than the missing values which is reflected in the count

In [391]:
population.describe()

country   1995   1996   1997   1998   1999   2000   2001   2002  \
count           197    197    197    197    197    197    197    196    196   
unique          197    184    182    180    184    183    188    186    182   
top     Afghanistan  10.1M  10.1M  10.9M  8.06M  10.2M  10.3M  10.2M  10.1M   
freq              1      3      4      3      2      4      3      4      3   

         2003   2004   2005   2006   2007   2008 2009   2010  
count     196    196    196    196    196    196  196    196  
unique    184    182    189    182    185    187  187    184  
top     22.6M  1.36M  10.5M  19.4M  10.6M  27.7M  10M  16.6M  
freq        2      2      3      3      2      2    3      3

In [392]:
population.dtypes

country    object
1995       object
1996       object
1997       object
1998       object
1999       object
2000       object
2001       object
2002       object
2003       object
2004       object
2005       object
2006       object
2007       object
2008       object
2009       object
2010       object
dtype: object

## 4. Fix the errors you identified in steps 2-3.

In [393]:
population.isna().sum()

country    0
1995       0
1996       0
1997       0
1998       0
1999       0
2000       0
2001       1
2002       1
2003       1
2004       1
2005       1
2006       1
2007       1
2008       1
2009       1
2010       1
dtype: int64

After inspecting the dataset for the missing values, we can see that all the missing values come from a single country, so it's safe to go ahead and drop the single row.

In [394]:
population.dropna(inplace=True)

In [395]:
gov_health_spending.isna().sum()

country    0
1995       3
1996       2
1997       2
1998       1
1999       1
2000       1
2001       1
2002       2
2003       2
2004       2
2005       2
2006       2
2007       2
2008       2
2009       2
2010       5
dtype: int64

We need to fix the missing values in our government health spending data. Our Government Health Spending dataset had significantly more missing values, but it's still less than 10% of our data, so we'll go ahead with these changes.

In [396]:
gov_health_spending.dropna(inplace=True)

In [397]:
gov_health_spending.shape

(184, 17)

## 5. Display the data types of each feature.

In [398]:
population.dtypes

country    object
1995       object
1996       object
1997       object
1998       object
1999       object
2000       object
2001       object
2002       object
2003       object
2004       object
2005       object
2006       object
2007       object
2008       object
2009       object
2010       object
dtype: object

In [399]:
life_exp.dtypes

country     object
1995       float64
1996       float64
1997       float64
1998       float64
1999       float64
2000       float64
2001       float64
2002       float64
2003       float64
2004       float64
2005       float64
2006       float64
2007       float64
2008       float64
2009       float64
2010       float64
dtype: object

In [400]:
gov_health_spending.dtypes

country     object
1995       float64
1996       float64
1997       float64
1998       float64
1999       float64
2000       float64
2001       float64
2002       float64
2003       float64
2004       float64
2005       float64
2006       float64
2007       float64
2008       float64
2009       float64
2010       float64
dtype: object

## 6. Fix any incorrect data types found in step 5.

Next I need to convert the values in my population dataset to numbers so I can explore it more thoroughly.

In [401]:
# Defining a function to convert my columns

letter_numbers = {'k':'000', 'm':'000000', 'b':'000000000'}

def convert_to_num(num):
    string = ''
    for char in num:
        if char.lower() in letter_numbers:
            string += letter_numbers[char.lower()]
        elif char in '1234567890.':
            string += char
    return float(string)

print(convert_to_num('12.1k'))

12.1


In [402]:
letter_numbers = {'k':1000, 'm':1000000, 'b':1000000000}

def convert_to_num(num):
    string = ''
    multiplier = 1
    for char in num:
        if char in '1234567890.':
            string += char
        elif char.lower() in letter_numbers:
            multiplier *= letter_numbers[char.lower()]
    result = float(string) * multiplier   
    return result

print(convert_to_num('12.1K'))

12100.0


In [403]:
for column in population.columns[1:]:
    population[column] = population[column].astype(str)

In [404]:
for column in population.columns[1:]:
    population[column] = population[column].apply(convert_to_num)

The columns are now all floats!

In [405]:
population.head()

country        1995        1996        1997        1998  \
0           Afghanistan  16400000.0  17100000.0  17800000.0  18500000.0   
1                Angola  13900000.0  14400000.0  14900000.0  15400000.0   
2               Albania   3280000.0   3270000.0   3250000.0   3230000.0   
3               Andorra     62900.0     64100.0     64700.0     65200.0   
4  United Arab Emirates   2430000.0   2570000.0   2750000.0   2920000.0   

         1999        2000        2001        2002        2003        2004  \
0  19300000.0  19500000.0  19700000.0  21000000.0  22600000.0  23600000.0   
1  15900000.0  16400000.0  16900000.0  17500000.0  18100000.0  18800000.0   
2   3210000.0   3180000.0   3150000.0   3120000.0   3090000.0   3060000.0   
3     65700.0     66100.0     67800.0     70800.0     73900.0     76900.0   
4   3100000.0   3280000.0   3450000.0   3630000.0   3810000.0   3990000.0   

         2005        2006        2007        2008        2009        2010  
0  24400000.0  25400000.0  25900000.0  26400000.0  27400000.0  28200000.0  
1  19500000.0  20200000.0  20900000.0  21700000.0  22500000.0  23400000.0  
2   3030000.0   3000000.0   2980000.0   2950000.0   2930000.0   2910000.0  
3     79800.0     80200.0     78200.0     76100.0     73900.0     71500.0  
4   4280000.0   4900000.0   5870000.0   6990000.0   7990000.0   8480000.0

In [406]:
gov_health_spending.set_index('country', inplace=True)

In [407]:
life_exp.set_index('country', inplace=True)

In [408]:
population.set_index('country', inplace=True)

In [409]:
population

1995        1996        1997        1998  \
country                                                                
Afghanistan           16400000.0  17100000.0  17800000.0  18500000.0   
Angola                13900000.0  14400000.0  14900000.0  15400000.0   
Albania                3280000.0   3270000.0   3250000.0   3230000.0   
Andorra                  62900.0     64100.0     64700.0     65200.0   
United Arab Emirates   2430000.0   2570000.0   2750000.0   2920000.0   
...                          ...         ...         ...         ...   
Samoa                   175000.0    177000.0    179000.0    180000.0   
Yemen                 16100000.0  16600000.0  17100000.0  17600000.0   
South Africa          44000000.0  44700000.0  45300000.0  45900000.0   
Zambia                 8680000.0   8900000.0   9130000.0   9370000.0   
Zimbabwe              11000000.0  11200000.0  11400000.0  11500000.0   

                            1999        2000        2001        2002  \
country                                                                
Afghanistan           19300000.0  19500000.0  19700000.0  21000000.0   
Angola                15900000.0  16400000.0  16900000.0  17500000.0   
Albania                3210000.0   3180000.0   3150000.0   3120000.0   
Andorra                  65700.0     66100.0     67800.0     70800.0   
United Arab Emirates   3100000.0   3280000.0   3450000.0   3630000.0   
...                          ...         ...         ...         ...   
Samoa                   182000.0    184000.0    186000.0    187000.0   
Yemen                 18100000.0  18600000.0  19100000.0  19700000.0   
South Africa          46400000.0  46800000.0  47200000.0  47700000.0   
Zambia                 9620000.0   9890000.0  10200000.0  10500000.0   
Zimbabwe              11700000.0  11800000.0  11900000.0  12000000.0   

                            2003        2004        2005        2006  \
country                                                                
Afghanistan           22600000.0  23600000.0  24400000.0  25400000.0   
Angola                18100000.0  18800000.0  19500000.0  20200000.0   
Albania                3090000.0   3060000.0   3030000.0   3000000.0   
Andorra                  73900.0     76900.0     79800.0     80200.0   
United Arab Emirates   3810000.0   3990000.0   4280000.0   4900000.0   
...                          ...         ...         ...         ...   
Samoa                   187000.0    188000.0    189000.0    189000.0   
Yemen                 20200000.0  20700000.0  21300000.0  22000000.0   
South Africa          48100000.0  48600000.0  49000000.0  49500000.0   
Zambia                10800000.0  11200000.0  11600000.0  12000000.0   
Zimbabwe              12100000.0  12200000.0  12200000.0  12300000.0   

                            2007        2008        2009        2010  
country                                                               
Afghanistan           25900000.0  26400000.0  27400000.0  28200000.0  
Angola                20900000.0  21700000.0  22500000.0  23400000.0  
Albania                2980000.0   2950000.0   2930000.0   2910000.0  
Andorra                  78200.0     76100.0     73900.0     71500.0  
United Arab Emirates   5870000.0   6990000.0   7990000.0   8480000.0  
...                          ...         ...         ...         ...  
Samoa                   190000.0    192000.0    193000.0    195000.0  
Yemen                 22600000.0  23300000.0  24000000.0  24700000.0  
South Africa          50000000.0  50600000.0  51200000.0  51800000.0  
Zambia                12400000.0  12900000.0  13300000.0  13800000.0  
Zimbabwe              12500000.0  12600000.0  12700000.0  12800000.0  

[196 rows x 16 columns]

## 7. Rename Columns.

In [411]:
pop_renamed = {col: 'pop' + col for col in population.columns}
population = population.rename(columns=pop_renamed)

In [412]:
life_exp_renamed = {col: 'life' + col for col in life_exp.columns}
life_exp = life_exp.rename(columns=life_exp_renamed)

In [413]:
gov_renamed = {col: 'gov' + col for col in gov_health_spending.columns}
gov_health_spending = gov_health_spending.rename(columns=gov_renamed)

## 8. Drop unnecessary rows (if needed).

We've already dropped all unnessecary rows other than countries that aren't shared between all our datasets, which we'll fix in the next step.

## 9. Merge dataframes that can be merged.

In [414]:
pop_and_life = pd.merge(population, life_exp, left_index=True, right_index=True)

In [419]:
pop_and_life.head()

pop1995     pop1996     pop1997     pop1998  \
country                                                                
Afghanistan           16400000.0  17100000.0  17800000.0  18500000.0   
Angola                13900000.0  14400000.0  14900000.0  15400000.0   
Albania                3280000.0   3270000.0   3250000.0   3230000.0   
Andorra                  62900.0     64100.0     64700.0     65200.0   
United Arab Emirates   2430000.0   2570000.0   2750000.0   2920000.0   

                         pop1999     pop2000     pop2001     pop2002  \
country                                                                
Afghanistan           19300000.0  19500000.0  19700000.0  21000000.0   
Angola                15900000.0  16400000.0  16900000.0  17500000.0   
Albania                3210000.0   3180000.0   3150000.0   3120000.0   
Andorra                  65700.0     66100.0     67800.0     70800.0   
United Arab Emirates   3100000.0   3280000.0   3450000.0   3630000.0   

                         pop2003     pop2004  ...  life2001  life2002  \
country                                       ...                       
Afghanistan           22600000.0  23600000.0  ...      54.8      55.5   
Angola                18100000.0  18800000.0  ...      53.4      54.5   
Albania                3090000.0   3060000.0  ...      76.0      75.9   
Andorra                  73900.0     76900.0  ...      80.9      81.1   
United Arab Emirates   3810000.0   3990000.0  ...      69.2      69.4   

                      life2003  life2004  life2005  life2006  life2007  \
country                                                                  
Afghanistan               56.5      57.1      57.6      58.0      58.5   
Angola                    55.1      55.5      56.4      57.0      58.0   
Albania                   75.6      75.8      76.2      76.9      77.5   
Andorra                   81.2      81.3      81.4      81.5      81.7   
United Arab Emirates      69.3      69.1      69.2      69.5      70.0   

                      life2008  life2009  life2010  
country                                             
Afghanistan               59.2      59.9      60.5  
Angola                    58.8      59.5      60.2  
Albania                   77.6      78.0      78.1  
Andorra                   81.8      81.8      81.8  
United Arab Emirates      70.4      70.6      70.8  

[5 rows x 32 columns]

In [420]:
country_data = pd.merge(pop_and_life, gov_health_spending, left_index=True, right_index=True)

In [421]:
country_data.head()

pop1995     pop1996     pop1997     pop1998     pop1999  \
country                                                                 
Angola     13900000.0  14400000.0  14900000.0  15400000.0  15900000.0   
Albania     3280000.0   3270000.0   3250000.0   3230000.0   3210000.0   
Andorra       62900.0     64100.0     64700.0     65200.0     65700.0   
Argentina  34900000.0  35400000.0  35800000.0  36200000.0  36700000.0   
Armenia     3320000.0   3300000.0   3270000.0   3240000.0   3210000.0   

              pop2000     pop2001     pop2002     pop2003     pop2004  ...  \
country                                                                ...   
Angola     16400000.0  16900000.0  17500000.0  18100000.0  18800000.0  ...   
Albania     3180000.0   3150000.0   3120000.0   3090000.0   3060000.0  ...   
Andorra       66100.0     67800.0     70800.0     73900.0     76900.0  ...   
Argentina  37100000.0  37500000.0  37900000.0  38300000.0  38700000.0  ...   
Armenia     3170000.0   3130000.0   3110000.0   3080000.0   3070000.0  ...   

           gov2001  gov2002  gov2003  gov2004  gov2005  gov2006  gov2007  \
country                                                                    
Angola        6.06     3.74     4.83     4.12     4.38     6.06     5.75   
Albania       7.24     7.32     7.64     9.23     9.79     9.05     8.46   
Andorra      19.20    20.00    22.00    22.70    22.00    22.80    21.30   
Argentina    14.30    15.30    14.80    15.20    14.30    14.40    13.90   
Armenia       6.73     6.19     6.76     6.76     6.79     7.42     6.81   

           gov2008  gov2009  gov2010  
country                               
Angola        6.40    10.10     7.18  
Albania       8.21     8.42     8.42  
Andorra      21.30    21.30    21.30  
Argentina    13.80    14.70    14.70  
Armenia       7.24     6.62     6.42  

[5 rows x 48 columns]

## 10. Perform any additional cleaning that you feel is necessary.

Using my source on which countries have universal healthcare, I'm going to add a column to my dataset reflecting that.

## 11. Save your cleaned and merged dataframes as csv files.

In [ ]:
population.to_csv('population_cleaned.csv')
life_exp.to_csv('life_exp_clean.csv')
gov_health_spending.to_csv('gov_health_spending_clean.csv')

### Data Dictionary

Now that we've fixed our data, and given it appropriate names, let's create a [data dictionary](http://library.ucmerced.edu/node/10249). 

A data dictionary provides a quick overview of features/variables/columns, alongside data types and descriptions. The more descriptive you can be, the more useful this document is.

Example of a Fictional Data Dictionary Entry: 

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**county_pop**|*integer*|2010 census|The population of the county (units in thousands, where 2.5 represents 2500 people).| 
|**per_poverty**|*float*|2010 census|The percent of the county over the age of 18 living below the 200% of official US poverty rate (units percent to two decimal places 98.10 means 98.1%)|

[Here's a quick link to a short guide for formatting markdown in Jupyter notebooks](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html).

Provided is the skeleton for formatting a markdown table, with columns headers that will help you create a data dictionary to quickly summarize your data, as well as some examples. **This would be a great thing to copy and paste into your custom README for this project.**

*Note*: if you are unsure of what a feature is, check the source of the data! This can be found in the README.

**To-Do:** *Edit the table below to create your own data dictionary for the datasets you chose.*

|Feature|Type|Dataset|Description|
|---|---|---|---|
|column name|int/float/object|This is an example| 


## Exploratory Data Analysis

Complete the following steps to explore your data. You are welcome to do more EDA than the steps outlined here as you feel necessary:
1. Summary Statistics.
2. Use a **dictionary comprehension** to apply the standard deviation function you create in part 1 to each numeric column in one of your dataframes.  **No loops**.
    - Assign the output to variable `sd` as a dictionary where: 
        - Each column name is now a key 
        - That standard deviation of the column is the value
        - *Example Output :* `{'1801': 3.80, '1802': 3.93, ...}`
3. Investigate trends in the data.
    - Using sorting and/or masking (along with the `.head()` method to avoid printing our entire dataframe), consider questions relevant to your problem statement. Some examples are provided below (but feel free to change these questions for your specific problem):
        - Which states have the highest and lowest populations for the past few decades?
        - Which states have the highest and lowest economic growth over the course of the dataset?
        - What years were there new countries created?
        - Do any countries have higher growth rates of population than their economy?
        - Which countries have had the lowest population growth?
    - **You should comment on your findings at each step in a markdown cell below your code block**. Make sure you include at least one example of sorting your dataframe by a column, and one example of using boolean filtering (i.e., masking) to select a subset of the dataframe.

#### 1. Summary Statistics

In [ ]:
population.describe()

In [ ]:
life_exp.describe

In [ ]:
gov_health_spending.describe()

2. Use a **dictionary comprehension** to apply the standard deviation function you create in part 1 to each numeric column in one of your dataframes.  **No loops**.
    - Assign the output to variable `sd` as a dictionary where: 
        - Each column name is now a key 
        - That standard deviation of the column is the value
        - *Example Output :* `{'1801': 3.80, '1802': 3.93, ...}`

In [ ]:
sd = {column : calc_std_dev(population[column]) for column in population.columns}

In [ ]:
sd

In [ ]:
std = {column : calc_std_dev(life_exp[column]) for column in life_exp.columns}

In [ ]:
std

3. Investigate trends in the data.
    - Using sorting and/or masking (along with the `.head()` method to avoid printing our entire dataframe), consider questions relevant to your problem statement. Some examples are provided below (but feel free to change these questions for your specific problem):
        - Which states have the highest and lowest populations for the past few decades?
        - Which states have the highest and lowest economic growth over the course of the dataset?
        - What years were there new countries created?
        - Do any countries have higher growth rates of population than their economy?
        - Which countries have had the lowest population growth?
    - **You should comment on your findings at each step in a markdown cell below your code block**. Make sure you include at least one example of sorting your dataframe by a column, and one example of using boolean filtering (i.e., masking) to select a subset of the dataframe.

In [ ]:
inner_merged_total = pd.merge(
...     population, life_exp, on=["1995", "2010"])

In [ ]:
inner_merged_total

**To-Do:** *Edit this cell with your findings on trends in the data (step 3 above).*

## Visualize the Data

There's not a magic bullet recommendation for the right number of plots to understand a given dataset, but visualizing your data is *always* a good idea. Not only does it allow you to quickly convey your findings (even if you have a non-technical audience), it will often reveal trends in your data that escaped you when you were looking only at numbers. It is important to not only create visualizations, but to **interpret your visualizations** as well.

**Every plot should**:
- Have a title
- Have axis labels
- Have appropriate tick labels
- Text is legible in a plot
- Plots demonstrate meaningful and valid relationships
- Have an interpretation to aid understanding

Here is an example of what your plots should look like following the above guidelines. Note that while the content of this example is unrelated, the principles of visualization hold:

![](https://snag.gy/hCBR1U.jpg)
*Interpretation: The above image shows that as we increase our spending on advertising, our sales numbers also tend to increase. There is a positive correlation between advertising spending and sales.*

---

Here are some prompts to get you started with visualizations. Feel free to add additional visualizations as you see fit:
1. Use Seaborn's heatmap with pandas `.corr()` to visualize correlations between all numeric features.
    - Heatmaps are generally not appropriate for presentations, and should often be excluded from reports as they can be visually overwhelming. **However**, they can be extremely useful in identify relationships of potential interest (as well as identifying potential collinearity before modeling).
    - Please take time to format your output, adding a title. Look through some of the additional arguments and options. (Axis labels aren't really necessary, as long as the title is informative).
2. Visualize distributions using histograms. If you have a lot, consider writing a custom function and use subplots.
    - *OPTIONAL*: Summarize the underlying distributions of your features (in words & statistics)
         - Be thorough in your verbal description of these distributions.
         - Be sure to back up these summaries with statistics.
         - We generally assume that data we sample from a population will be normally distributed. Do we observe this trend? Explain your answers for each distribution and how you think this will affect estimates made from these data.
3. Plot and interpret boxplots. 
    - Boxplots demonstrate central tendency and spread in variables. In a certain sense, these are somewhat redundant with histograms, but you may be better able to identify clear outliers or differences in IQR, etc.
    - Multiple values can be plotted to a single boxplot as long as they are of the same relative scale (meaning they have similar min/max values).
    - Each boxplot should:
        - Only include variables of a similar scale
        - Have clear labels for each variable
        - Have appropriate titles and labels
4. Plot and interpret scatter plots to view relationships between features. Feel free to write a custom function, and subplot if you'd like. Functions save both time and space.
    - Your plots should have:
        - Two clearly labeled axes
        - A proper title
        - Colors and symbols that are clear and unmistakable
5. Additional plots of your choosing.
    - Are there any additional trends or relationships you haven't explored? Was there something interesting you saw that you'd like to dive further into? It's likely that there are a few more plots you might want to generate to support your narrative and recommendations that you are building toward. **As always, make sure you're interpreting your plots as you go**.

In [423]:
continent_mapping = {
    'Africa': ['Algeria', 'Angola', 'Egypt', 'Nigeria', 'South Africa', 'Kenya', 'Morocco'],
    'Asia': ['China', 'India', 'Indonesia', 'Japan', 'South Korea', 'Saudi Arabia', 'Turkey'],
    'Europe': ['France', 'Germany', 'Italy', 'Russia', 'Spain', 'United Kingdom'],
    'North America': ['Canada', 'Mexico', 'United States'],
    'Oceania': ['Australia', 'New Zealand'],
    'South America': ['Argentina', 'Brazil', 'Chile', 'Colombia', 'Peru']
}

In [ ]:
# Code

## Conclusions and Recommendations

Based on your exploration of the data, what are you key takeaways and recommendations? Make sure to answer your question of interest or address your problem statement here.

**To-Do:** *Edit this cell with your conclusions and recommendations.*

**Don't forget to create your README!**

**To-Do:** *If you combine your problem statement, data dictionary, brief summary of your analysis, and conclusions/recommendations, you have an amazing README.md file that quickly aligns your audience to the contents of your project.* Don't forget to cite your data sources!